Add registered MOBJs

lorc opened the pull request:
Those memory objects represent registered memory buffers. This PR is part of #1232

jenswi-linaro commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -766,7 +824,17 @@ int core_va2pa_helper(void *va, paddr_t *pa)
+	 * together with an invalid null physical address.
+	 */
+	if (map->pa)
+		*pa = ((vaddr_t)va & (map->region_size - 1)) |
```
These line should be
```
*pa = map->pa + (vaddr_t)va  - map->va;
```
now

etienne-lms commented on core/arch/arm/mm/mobj.c:
```diff
@@ -312,6 +308,291 @@ struct mobj *mobj_mm_alloc(struct mobj *mobj_parent, size_t size,
+	struct mobj mobj;
+	tee_mm_entry_t *mm_entry;
+	struct mobj *reg_shm;
+	bool owns_reg_shm;
```
is this `owns_reg_shm` necessary ? a non-null `reg_shm` already indicates a registered reg_shm mobj is realted to it.

lorc replied:
Ah yes. This is leftover from earlier version. There was to constructors for `mobj_mapped_shm`. One constructor created `mobj` from a list of pages, and other created from already existing `mobj_registered_shm`. So I had to distinguish those two cases. Now it is redundant.

lorc commented on core/arch/arm/include/mm/mobj.h:
```diff
@@ -39,6 +39,7 @@
 struct mobj {
 	const struct mobj_ops *ops;
 	size_t size;
+	size_t granule;
```
Currently it is used only in `set_pg_region()` to make it map regions page-by-page. This is crucial for registered buffers. It means something like "longest physically contiguous part of the mobj".   
So, in case of registered mobj, `granule` will be `SMALL_PAGE_SIZE`.  Probably, this is not the best solution. But, I have no idea how to do this better. So any suggestions will be appreciated :)

etienne-lms replied:
Ok, thanks. Maybe it could be renamed `phys_granule` for clarity. 

Since this granule is fixed for each 'kind' of memory object, I would prefer this to be provided through the related object operations, Either through a function, or straight from a size_t value in the `struct mobjs_ops`.

lorc replied:
Okay, `phys_granule` is better name. There are `mobj_get_granule()` function that returns corresponding value.

etienne-lms commented on core/arch/arm/mm/mobj.c:
```diff
@@ -155,12 +157,6 @@ struct mobj *mobj_phys_alloc(paddr_t pa, size_t size, uint32_t cattr,
 	enum teecore_memtypes area_type;
 	void *va;
 
-	if ((pa & CORE_MMU_USER_PARAM_MASK) ||
```
I tried to find back why this alignment constraint was checked. @jenswi-linaro, do you see any reason why this should be preserved, or maybe moved to other specific places for already implement mobj allocation requests?

jenswi-linaro replied:
Now that you mention it, I don't see why it should be removed. From what I've seen so far is the dynamic shared memory stuff depending on `CFG_SMALL_PAGE_USER_TA` which is also defining `CORE_MMU_USER_PARAM_SHIFT` as `SMALL_PAGE_SHIFT`.

lorc replied:
@jenswi-linaro I use `mobj_phys` to represent buffer in static SHM. This buffers are not page-aligned. So, I either should remove this check, or add another `mobj` type.

jenswi-linaro replied:
I think I'd rather have another `mobj` type. With that we could use the same physical `mobj` for a references into the shared memory pool.

Another option is to return an offset that should be used with the returned `mobj` from `thread_rpc_alloc()` also, but I think it's clearer with another `mobj` type that hides this.

lorc replied:
Okay, then I returned back this check in v2 delta. 
I'll add the new `mobj` type in different PR if you don't mind.

jenswi-linaro commented:
FYI, I've created https://bugs.linaro.org/show_bug.cgi?id=3006 to deal with the AArch64 linking error. I'm using one of the patches in this PR as an example.

lorc commented:
@jenswi-linaro, thank you!

lorc commented:
According to Adhemerval Zanella, this linker error is caused by `--gc-sections- flag`. With this small patch problem vanishes and I don't see any difference in resulting binary size:

```
diff --git a/core/arch/arm/kernel/link.mk b/core/arch/arm/kernel/link.mk
index 4a7bd8e..49eb7a5 100644
--- a/core/arch/arm/kernel/link.mk
+++ b/core/arch/arm/kernel/link.mk
@@ -82,7 +82,7 @@ entries-init += _start
 objs-init := \
 	$(filter-out $(addprefix $(out-dir)/, $(objs-init-rem)), $(objs) \
 		$(link-out-dir)/version.o)
-ldargs-init := -i --gc-sections $(addprefix -u, $(entries-init))
+ldargs-init := -i $(addprefix -u, $(entries-init))
 
 ldargs-init-objs := $(objs-init) $(link-ldadd) $(libgcccore)
 cleanfiles += $(link-out-dir)/init.o
```
 
Should we wait for proper fix from compiler team or should we apply this workaround?

jenswi-linaro commented:
I have a fix in https://github.com/OP-TEE/optee_os/pull/1562

lorc commented:
I have rebased this PR onto current master to get #1562 in history. Also I have added v3 delta patch that updates `link_dummy.ld`.

jenswi-linaro commented on core/arch/arm/kernel/link_dummy.ld:
```diff
@@ -46,6 +46,8 @@ __early_bss_start = .;
 __end = .;
 __end_phys_mem_map_section = .;
 __end_phys_sdp_mem_section = .;
+__start_phys_nsec_ddr_section = .;
```
Please keep this list sorted as the command `sort` would sort it.

lorc commented:
Okay, entries in `link_dummy.ld` are  sorted in v4 patch.

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

lorc commented:
Thank you, Jens. 
I have squashed deltas and added `RB` tag. Travis build fails due to `checkpatch` warnings. Some of them are `checkpatch` bugs, and some are real, but they are needed (`extern` variables).

jenswi-linaro commented:
@etienne-lms are you OK with this too?

etienne-lms commented:
yes, looks ok to me.
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

Few comments that could be addressed in later changes:

- In **core_mmu.c** `init_mem_map()`, i would propose that static mappings asserts i wont map a null physical address.
```c
	/* Assert static map do not mapped to a null phys address */
	assert(map->pa || core_mmu_is_dynamic_vaspace(map));
```

- Regarding the `extern` checkpatch issue, I think a later change could move all these `extern` into some local **core_mmu_private.h** header file.

lorc commented:
@etienne-lms, thank you for review. I had moved `extern` declarations to `linker.h` file which was added recently.
I added Etienne's RB tag and rebased this on current master branch.

jforissier commented:
@lorc it looks like we're missing Etienne's R-b.

lorc commented:
I'm sorry. I lost them during rebasing to master branch. Should be fine now.

jforissier commented:
No worries.
One more thing: `CFG_SMALL_PAGE_USER_TA=n` has been removed recently so there are some useless lines. Do you want to fix that? Or we can do it in another patch, not a problem for me.

lorc commented:
Oh, this a good news. I'll remove redundant code now, if you don't mind.

jforissier commented:
Sure.

lorc commented:
Done. It was only in `mobj.c` and `mobj.h`.
